app/models/git_repository.rb

```diff
@@ -26,91 +27,135 @@ def refresh
   end
 
   def branch_info(name)
-    refresh unless repo
+    ensure_refreshed
     branch = git_branches.detect { |item| item.name == name }
     raise "Branch #{name} not found" unless branch
     branch.attributes.slice(*INFO_KEYS)
   end
 
   def tag_info(name)
-    refresh unless repo
+    ensure_refreshed
     tag = git_tags.detect { |item| item.name == name }
     raise "Tag #{name} not found" unless tag
     tag.attributes.slice(*INFO_KEYS)
   end
 
+  def entries(ref, path)
+    ensure_refreshed
+    with_worktree do |worktree|
+      worktree.ref = ref
+      worktree.entries(path)
+    end
+  end
+
+  def checkout(ref, path)
+    ensure_refreshed
+    with_worktree do |worktree|
+      message = "Checking out #{url}@#{ref} to #{path}..."
+      _log.info(message)
+      worktree.ref = ref
+      worktree.checkout(path)
+      _log.info("#{message}...Complete")
+    end
+    path
+  end
+
   def directory_name
-    parsed = URI.parse(url)
-    raise "Invalid URL missing path" if parsed.path.blank?
-    File.join(GIT_REPO_DIRECTORY, parsed.path)
+    raise ActiveRecord::RecordNotSaved if new_record?
```

Comment:
Honestly, no one should hit this error...it's more of a catch just in case someone does it in Rails console or something, and by then the persona is a developer, so ActiveRecord::RecordNotSaved should be clear enough on how to "fix" it.